sacrebleu.py

```diff
@@ -1461,6 +1516,59 @@ def get_a_list_of_testset_names():
     return message
 
 
+def _available_origlangs(test_sets, langpair):
+    """Return a list of origlang values in according to the raw SGM files."""
+    origlangs = set()
+    for test_set in test_sets.split(','):
+        rawfile = os.path.join(SACREBLEU_DIR, test_set, 'raw', DATASETS[test_set][langpair][0])
+        if rawfile.endswith('.sgm'):
+            with smart_open(rawfile) as fin:
+                for line in fin:
+                    if line.startswith('<doc '):
+                        doc_origlang = re.sub(r'.* origlang="([^"]+)".*\n', '\\1', line)
+                        origlangs.add(doc_origlang)
+    return sorted(list(origlangs))
+
+
+def _filter_subset(systems, test_sets, langpair, origlang, subset=None):
+    """Filter sentences with a given origlang (or subset) according to the raw SGM files."""
+    if origlang is None and subset is None:
+        return systems
+    if test_sets is None or langpair is None:
+        raise ValueError('Filtering for --origlang or --subset needs a test (-t) and a language pair (-l).')
+
+    indices_to_keep = []
+    for test_set in test_sets.split(','):
+        rawfile = os.path.join(SACREBLEU_DIR, test_set, 'raw', DATASETS[test_set][langpair][0])
+        if not rawfile.endswith('.sgm'):
+            raise Exception('--origlang and --subset supports only *.sgm files, not %s', rawfile)
+        if subset is not None:
+            if test_set not in SUBSETS:
+                raise Exception('No subset annotation available for test set ' + test_set)
+            doc_to_tags = SUBSETS[test_set]
+        number_sentences_included = 0
+        with smart_open(rawfile) as fin:
+            include_doc = False
+            for line in fin:
+                if line.startswith('<doc '):
+                    if origlang is None:
+                        include_doc = True
+                    else:
+                        doc_origlang = re.sub(r'.* origlang="([^"]+)".*\n', '\\1', line)
+                        if origlang.startswith('non-'):
+                            include_doc = doc_origlang != origlang[4:]
+                        else:
+                            include_doc = doc_origlang == origlang
+                    if subset is not None:
+                        doc_id = re.sub(r'.* docid="([^"]+)".*\n', '\\1', line)
+                        if not re.search(subset, doc_to_tags.get(doc_id, '')):
+                            include_doc = False
+                if line.startswith('<seg '):
+                    indices_to_keep.append(include_doc)
+                    number_sentences_included += 1 if include_doc else 0
+    return [[sentence for sentence,keep in zip(sys, indices_to_keep) if keep] for sys in systems]
+
+
 def main():
```

Comment:
This would add some work: but I've thought for a while that we should get rid of this hacky XML parsing in favor of using `lxml.etree`. This would simplify a lot of code and remove a potential for errors to creep in. What do you think? We use this library extensively in the Anthology.

Reply:
I would prefer to merge this PR first and possibly do the lxml refactoring in a separate PR, after considering the following:
- I used the ad-hoc reg-ex style, which was already there for the SGM->plaintext conversion. It's just a few lines of code and I doubt the lxml refactoring can be notably shorter/simpler (though I am not familiar with lxml).
- I thought that sacreBLEU is meant to be usable also as a single script with no installation, i.e. no dependencies (meanwhile, `portalocker` was added, I know). That said, if dependencies are allowed now, we could reconsider re-implementing everything with numpy and adding bootstrap resampling (#11) and paired bootstrap tests.
- Are you sure that all legacy WMT *.sgm files are XML 1.0 valid and `lxml.etree` parses them correctly? They have no `<?xml version="1.1" encoding="UTF-8"?>` declaration, so they are not valid XML 1.1. Some older testsets (wmt08, I think) use `<DOC>` instead of `<doc>` etc.

Reply:
I'm not sure, but my impression is that `lxml.etree` is flexible against imperfect specs.

I don't mind if sacrebleu has dependencies, since `pip` makes it easy to install them and doesn't IMO make it harder to use. Rewriting BLEU in `numpy` would be *awesome*. It's too slow as it is. Did you have a reference `numpy` implementation in mind? I hadn't given any thought about how to make this faster.